store/iterator.go

```diff
@@ -108,163 +95,158 @@ func (b *btreeIter) wrap(parent Iterator) *itemIter {
 		wrap:   b,
 		parent: parent,
 	}
-	if err := iter.skipAllDeleted(); err != nil {
-		panic(err)
-	}
 	return iter
 }
 
-func (b *btreeIter) next() {
-	b.data, b.hasMore = <-b.read
+func (b *btreeIter) Next() (keyer, error) {
+	data, hasMore := <-b.read
+	if !hasMore {
+		return nil, errors.Wrap(errors.ErrDone, "btree iterator")
+	}
+	key, ok := data.(keyer)
+	if !ok {
+		return nil, errors.Wrapf(errors.ErrType, "expected keyer, got %T", data)
+	}
+	return key, nil
 }
 
-func (b *btreeIter) close() {
+func (b *btreeIter) Release() {
 	b.once.Do(func() {
 		b.stop <- struct{}{}
 	})
 }
 
-// get requires this is valid, gets what we are pointing at
-func (b *btreeIter) get() keyer {
-	return b.data.(keyer)
-}
-
-func (b *btreeIter) valid() bool {
-	return b.hasMore
-}
-
 type itemIter struct {
 	wrap *btreeIter
 	// if we are iterating in a cache-wrap (and who isn't),
 	// we need to combine this iterator with the parent
 	parent Iterator
+
+	parentDone   bool
+	cachedParent Model
+	wrapDone     bool
+	cachedWrap   keyer
```

Comment:
pp: `keyer` is an interface and you need to cast it in several places. Another option would be to use the `setItem` type here directly as it is required by L232 anyway and drop all the (other) casts.

Reply:
Sometimes it is a deleteItem, 
https://github.com/iov-one/weave/blob/cdc50e5550f30c74090144233197408c7a0c4c20/store/iterator.go#L214-L218
Is needed handling when we have a deleteItem covering up an item in the parent store

Reply:
that is satisfied by setItem, too